### PR TITLE
Inner blocks: edit/save symmetry + stabilise API

### LINKED ISSUE #35071: Get `useInnerBlocksProps` hook out of experimental status
### What problem does this address?

The `useInnerBlocksProps` hook was added in October 2020 via #25633 as an experimental hook. Since its release, a few bugs have appeared (#34120, #30408, #29513, #29482, #29036, #14242) but, in general, no breaking feedback has been received. However, this might be either because the hook version is very stable or because of a lack of wider testing.

### What is your proposed solution?

We should consider moving the `useInnerBlocksProps` out from experimental as the next step towards wider adoption and as a means to receive more organic feedback from users.

### Alternate options

If we feel it is not ready to be marked as a stable API I think we would benefit from adding some experimental documentation for plugin authors etc to test this hook and provide feedback.

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Previously #25644 and #25633.
Fixes #35071.

This Replaces the `InnerBlocks.Content` component with a function `useInnerBlocksProps.save` for the `save` component so it's similar in use as `useInnerBlocksProps` in the `edit` component. There's no need for an API version bump since both can be used, but the goal would be to eventually deprecate `InnerBlocks.Content`.

The only instances that I didn't convert were dynamic blocks that have save functions that return only `<innerBlocks.Content />`. I'm not sure what the reason is to return the inner blocks without a wrapper. I see some dynamic blocks that don't have a save function at all or that do have a wrapper div.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by fabiankaegy** [comment]
@ellatrix I think we should also update the documentation of the Component itself here: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/inner-blocks/README.md as part of this PR. 

I can take care of that once you rebased this pr if that works for you (y)

File: packages/block-editor/src/components/index.js [hunk 1]
```diff
@@ -54,7 +54,7 @@ export {
 export { default as __experimentalImageSizeControl } from './image-size-control';
 export {
 	default as InnerBlocks,
-	useInnerBlocksProps as __experimentalUseInnerBlocksProps,
+	useInnerBlocksProps as useInnerBlocksProps,
```
Comment (mkevins) [inline packages/block-editor/src/components/index.js]: NP: no need to alias now. :)

Comment (ellatrix) [reply packages/block-editor/src/components/index.js]: Oops. We should have the linter catch stuff like this.

Comment (mkevins) [reply packages/block-editor/src/components/index.js]: Good idea. I was surprised it didn't, actually, but I guess it isn't common :shrug:

**Review by fabiankaegy — COMMENTED** [review]
I have tested this locally with the latest changes and could not find any issues. I build a page that used the different blocks on the current trunk and loaded that page again with this PR active. And no changes were apparent. No errors were logged to the console

File: packages/block-library/src/gallery/save.js [hunk 2]
```diff
@@ -25,10 +29,12 @@ export default function saveWithInnerBlocks( { attributes } ) {
...
 		'is-cropped': imageCrop,
 	} );
+	const blockProps = useBlockProps.save( { className } );
+	const innerBlocksProps = useInnerBlocksProps.save( blockProps );
 
 	return (
-		<figure { ...useBlockProps.save( { className } ) }>
-			<InnerBlocks.Content />
+		<figure { ...innerBlocksProps }>
+			{ innerBlocksProps.children }
```
Comment (Mamaduka) [inline packages/block-library/src/gallery/save.js:37]: Maybe we should document this usage? When we want to render inner blocks' contents explicitly.

**Review by Mamaduka — APPROVED** [review]
The changes test well for me. I used a similar method of testing as @fabiankaegy and couldn't spot any errors or regressions.

> I'm not sure what the reason is to return the inner blocks without a wrapper.

Some dynamic blocks (mostly FSE) try not to render anything if data doesn't exist. This is the only reason I've seen for not using the wrapper element.

**Comment by fabiankaegy** [comment]
Thank you @ellatrix for all your work on this :) I really appreciate your work here and just want you to know that it is very much appreciated! Excited for this to be I. 5.9 :)

**Comment by fabiankaegy** [comment]
## Dev Note Draft: 

# Take more control over Inner Block Areas (as a block developer)

Prior to WordPress 5.9 the only way to work with inner block was to use the `<InnerBlocks />` component. One downside of this approach is that in the editor this creates additional DOM nodes that wrap the markup of the inner blocks. This makes it more difficult to style these inner blocks areas to match what the enduser will see on the site.

With WordPress 5.9 we're introducing a new react hook called `useInnerBlocksProps` that allows you to change this and take more control over the markup of inner blocks areas. The `useInnerBlocksProps` is exported from the `@wordpress/block-editor` package same as the `InnerBlocks` component itself and supports everything the component does. It also works like the `useBlockProps` hook [introduced with `apiVersion: 2` in WordPress 5.6](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-api-versions.md#version-2--wordpress-56).

To use the hook, take the object returned from calling the hook and spread it onto the host element you want to render the inner blocks into. For example:

```js
function BlockEdit(props) {
    const blockProps = useBlockProps();
    const innerBlocksProps = useInnerBlocksProps();

    return (
        <section {...blockProps}>
             <div {...innerBlocksProps} />
        </section>
    );
}
```

The above code will render to the following markup in the editor:

```html
<section>
    <div>
        <!-- Inner Blocks get inserted here -->
    </div>
</section>
```

The hook also accepts two arguments:

1. The first argument accepts an object containing additional props that are added to the element in the dom. For example, to add a custom class name to the inner blocks area you pass `{ className: 'my-class' }` as the first argument.
2. The second argument is also an object and can contain any properties to control the behavior of the inner blocks area. It accepts all options like `allowedBlocks`, `template`, `templateLock`, `renderAppender`, etc. You can find the full list of supported properties in the [InnerBlocks props documentation]( https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/inner-blocks#props)

```js
useInnerBlocksProps( additionalWrapperProps, innerBlocksProps );
```

## Saving the inner blocks

To save inner blocks content there is a special variant of the hook, this is similar to the component-based approach that uses `<InnerBlocks.Content />` in the `save` method of a block. For hooks, you add `.save` to the end of the hook like so:

```js

function BlockSave( props ) {
    const innerBlocksProps = useInnerBlocksProps.save( { className: 'my-class' } );

    return (
        <section {...innerBlocksProps} ) />
    );
}
```

When you use the `useInnerBlocksProps.save()` hook it only accepts the first parameter, these are the options you want to add to the container element. 

## Why use hooks?

The hooks allow for new options and nice enhancements. For example, you can take the object returned from the `useBlockProps` hook—these contain the properties for the block's wrapping container—and pass them to the `useInnerBlocksProps` hook. This reduces the number of elements we need to create. For example:

```js
function BlockEdit(props) {
    const blockProps = useBlockProps();
    const innerBlocksProps = useInnerBlocksProps( blockProps );

    return (
        <section {...innerBlocksProps} />
    );
}
```

```html
<section>
    <!-- Inner Blocks get inserted here -->
</section>
```

Another benefit to using the hook approach is using the returned value, which is just an object, and deconstruct to get the react `children` from the object. This property contains the actual child inner blocks thus we can place elements on the same level as our inner blocks. 

```js
function BlockEdit(props) {
    const blockProps = useBlockProps();
    const { children, ...innerBlocksProps } = useInnerBlocksProps( blockProps );

    return (
        <section {...innerBlocksProps}>
            { children }
            <!-- Insert any arbitrary html here at the same level as the children -->
        </section>
    );
}
```

```html
<section>
    <!-- Inner Blocks get inserted here -→
    <!-- The custom html gets rendered on the same level -->
</section>
```


## Example:

`block.json`
```json
{
    "apiVersion": 2,
    "name": "example/block",
    "title": "Example",
    "category": "text",
    "editorScript": "file:./index.js"
}
```

`index.js`
```js
import { registerBlockType } from '@wordpress/blocks';
import { useBlockProps, useInnerBlocksProps } from '@wordpress/block-editor';

import metadata from './block.json';

function BlockEdit(props) {
    const blockProps = useBlockProps( { className: 'my-class' } );
    const innerBlocksProps = useInnerBlocksProps(
        blockProps,
        { allowedBlocks: [ 'core/heading', 'core/paragraph', 'core/image' ] }
    );

    return (
        <section {...innerBlocksProps} />
    );
}

function BlockSave(props) {
    const blockProps = useBlockProps.save( { className: 'my-class' } );
    const innerBlocksProps = useInnerBlocksProps.save( blockProps );

    return <section {...innerBlocksProps} />
}

registerBlockType( metadata, {
    edit: BlockEdit,
    save: BlockSave,
} )
```
--- 

Tagging @gziolo @mkaz and @ellatrix for review

**Comment by gziolo** [comment]
@fabiankaegy, great working putting this dev note together. It is nicely structured as each section builds on top of each other and includes nicely crafted examples. I don't know the new API fully but I find this explanation very informative 💯 

One thing that would be worth explaining explicitly is that similar to how `InnerBlocks` and `InnerBlocks.Content` work, we have two different hooks for `edit` and `save`. I see that the last example covers it, but I almost missed that part.

**Comment by fabiankaegy** [comment]
@gziolo Updated the comment with a new section called `Saving the inner blocks`

**Comment by mkaz** [comment]
@fabiankaegy This looks good, thanks for taking the time to put it together. I edited the note above for clarity and wording in a couple spots (you can see a diff of the edits using the dropdown).

It is good to publish to make/core, use the `dev-notes` and `5-9` tags. 🚀 
Let me know if you have any questions. Thanks again! 🥇

**Comment by ellatrix** [comment]
> https://make.wordpress.org/core/2021/12/28/take-more-control-over-inner-block-areas-as-a-block-developer/

Thank you for the dev note @fabiankaegy! It looks great.

**Pull request merged**